gqcp/include/QCModel/HF/GHF.hpp

```diff
@@ -0,0 +1,395 @@
+// This file is part of GQCG-GQCP.
+//
+// Copyright (C) 2017-2020  the GQCG developers
+//
+// GQCG-GQCP is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// GQCG-GQCP is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with GQCG-GQCP.  If not, see <http://www.gnu.org/licenses/>.
+
+#pragma once
+
+
+#include "Basis/TransformationMatrix.hpp"
+#include "Operator/FirstQuantized/ElectronicSpinOperator.hpp"
+#include "Operator/SecondQuantized/SQHamiltonian.hpp"
+#include "Utilities/aliases.hpp"
+
+
+namespace GQCP {
+namespace QCModel {
+
+
+/**
+ *  The generalized Hartree-Fock wave function model.
+ * 
+ *  @tparam _Scalar             the type of scalar that is used for the expansion of the spatial orbitals in their underlying scalar basis
+ */
+template <typename _Scalar>
+class GHF {
+public:
+    using Scalar = _Scalar;
+
+
+private:
+    size_t N;  // the number of electrons
+
+    VectorX<double> orbital_energies;  // sorted in ascending energies
+    TransformationMatrix<Scalar> C;    // the coefficient matrix that expresses every spinor orbital (as a column) in the underlying scalar bases
+
+
+public:
+    /*
+     *  CONSTRUCTORS
+     */
+
+    /**
+     *  The standard member-wise constructor
+     * 
+     *  @param N                    the number of electrons
+     *  @param C                    the coefficient matrix that expresses every spinor orbital (as a column) in the underlying scalar bases
+     *  @param orbital_energies     the GHF MO energies
+     */
+    GHF(const size_t N, const VectorX<double>& orbital_energies, const TransformationMatrix<double>& C) :
+        N {N},
+        orbital_energies {orbital_energies},
+        C(C) {}
+
+
+    /**
+     *  Default constructor setting everything to zero
+     */
+    GHF() :
+        GHF(0, TransformationMatrix<double>::Zero(0, 0), VectorX<double>::Zero(0)) {}
```

Comment:
I think your members are switched. Plus add the `Scalar` type for the transformation matrix.

```suggestion
        GHF(0, VectorX<double>::Zero(0), TransformationMatrix<Scalar>::Zero(0, 0)) {}
```

Maybe we'll have to let the orbital energies be of type `VectorX<Scalar>`, even though we know that they should be real.